indexer.go

```diff
@@ -0,0 +1,862 @@
+	indexes, err := collection.Indexes(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var best *Index
+	bestMatchingFields := 0
+
+	for _, index := range indexes {
+		// Iterate over the fields
+		matchingFields := 0
+		for _, field := range index.fields {
+			_, ok := query.Equal[field]
+			if ok {
+				matchingFields += 1
+			}
+		}
+		if matchingFields > bestMatchingFields {
+			best = &index
+			bestMatchingFields = matchingFields
+		} else if matchingFields == bestMatchingFields {
+			if len(best.fields) > len(index.fields) {
+				best = &index
+			}
+		}
+	}
+
+	return best, nil
+}
+
+func (query Query) Matches(record Record) bool {
+	for field, expected := range query.Equal {
+		value, err := record.Data.LookupByString(field)
+
+		if err != nil || value == nil {
+			return false
+		}
+
+		if datamodel.DeepEqual(value, expected) != true {
+			return false
+		}
+	}
+
+	return true
+}
+
+func ParseStringsFromCBOR(data []byte) ([]string, error) {
+	ipldList, err := ParseListFromCBOR(data)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var items []string
+
+	iterator := ipldList.ListIterator()
+
+	for !iterator.Done() {
+		_, value, err := iterator.Next()
+
+		if err != nil {
+			return nil, err
+		}
+
+		asString, err := value.AsString()
+
+		if err != nil {
+			return nil, err
+		}
+
+		items = append(items, asString)
+
+	}
+
+	return items, nil
+}
+
+func ParseListFromCBOR(data []byte) (ipld.Node, error) {
+	builder := basicnode.Prototype.List.NewBuilder()
+	reader := bytes.NewReader(data)
+	err := dagcbor.Decode(builder, reader)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return builder.Build(), nil
+}
+
+func EncodeListToCBOR(data []ipld.Node) ([]byte, error) {
+	assembleKeyNode := func(am datamodel.ListAssembler) {
+		for _, value := range data {
+			qp.ListEntry(am, qp.Node(value))
+		}
+	}
+
+	keyNode, err := qp.BuildList(basicnode.Prototype.Any, -1, assembleKeyNode)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+
+	err = dagcbor.Encode(keyNode, &buf)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+func IndexKeyFromFields(fields []string) ([]byte, error) {
+	assembleKeyNode := func(am datamodel.ListAssembler) {
+		for _, key := range fields {
+			qp.ListEntry(am, qp.String(key))
+		}
+	}
+
+	keyNode, err := qp.BuildList(basicnode.Prototype.Any, -1, assembleKeyNode)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+
+	err = dagcbor.Encode(keyNode, &buf)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+func IndexKeyFromRecord(keys []string, record ipld.Node, id []byte) ([]byte, error) {
+	var hadError error
+	assembleKeyNode := func(am datamodel.ListAssembler) {
+		for _, key := range keys {
+			value, err := record.LookupByString(key)
+			if err != nil {
+				hadError = err
+				break
+			}
+			qp.ListEntry(am, qp.Node(value))
+		}
+		if id != nil && len(id) != 0 {
+			qp.ListEntry(am, qp.Bytes(id))
+		}
+	}
+
+	if hadError != nil {
```

Comment:
assembleKeyNode() is executed in qp.BuildList() below, so the hadError must be nil